File: Chapter1/Q1/Q1.py
# Solution1 : sort
def isUnique1(str):
    temp = sorted(str)
    for i in range(1,len(temp)):
        if temp[i]==temp[i-1]:
            return False
    return True

# Solution2 : ASCII Pool
def isUnique2(str):
    pool = [0 for i in range(128)]
    for i in range(len(str)):
        if pool[ord(str[i])]>0:
            return False
        else:
            pool[ord(str[i])]+=1
    return True

def isUnique3(str):
    pool = 0
    for i in range(len(str)):
        temp = ord(str[i])-ord('a')
        if (pool&(1<<temp))>0:
            return False
        else:
            pool|=1<<temp
    return True

a = "akddpol"
b = "zbcdet"
print isUnique3(a) , isUnique3(b)

File: Chapter1/Q1/Q1.java
import java.util.Arrays;

/* 
    Is Unique: Implement an algorithm to determine if a string has all unique characters.
    What if you cannot use additional data structures? 
*/

public class Q1{
    public static void main (String[] arg)
    {
        String a = "abcddop";
        String b = "abrtppo";
        System.out.println(isUnique3(a));
        System.out.println(isUnique3(b));

    }

    // Solution1 : Sort first
    public static boolean isUnique1(String str){
        char[] s = str.toCharArray();
        Arrays.sort(s);
        for(int i=1 ;i<s.length ; i++)
        {
            if(s[i]==s[i-1])
                return false;
        }      
        return true;
    }

    // Solution2 : ASCII pool
    public static boolean isUnique2(String str)
    {
        int[] pool = new int[128];
        for(char c : str.toCharArray())
        {
            if(pool[c]>0)
                return false;
            else
                pool[c]++;
        }
        return true;
    }

    // Solution 3 : Bit Manipulation (work for [a-z])
    public static boolean isUnique3(String str)
    {
        int pool = 0;
        for (char c : str.toCharArray())
        {
            int offset = c-'a';
            if((pool&(1<<offset))>0)
                return false;
            else
                pool|=1<<offset;
        }
        return true;
    }
}   